project_template/CMakeLists.txt

```diff
@@ -0,0 +1,104 @@
+# Copyright (c) Microsoft Corporation. All rights reserved.
+# SPDX-License-Identifier: MIT
+
+cmake_minimum_required(VERSION 3.15)
+
+project(az_exampleshortname
+    LANGUAGES C CXX
+    VERSION 0.1.0)
+
+option(az_exampleshortname_with_docs "Build documentation" ON)
+option(BUILD_TESTING "Build tests" ON)
+
+
+
+
+
+# this is a cmake 3.15 feature that tells find_package to 
+# look at config files before module files. This makes it easier
+# to write buildsystems that work on many platforms, espicially
+# when finding libraries that have multiple build systems.
+
+
+# if needed: add ./cmake to the module path
+list(APPEND CMAKE_MODULE_PATH ${CMAKE_CURRENT_LIST_DIR}/cmake)
+
+set(CMAKE_C_STANDARD 99)
+set(CMAKE_CXX_STANDARD 17)
+
+# default to hidden symbols when building shared libraries
+# on unix systems
+set(CMAKE_C_VISIBILITY_PRESET hidden)
+set(CMAKE_CXX_VISIBILITY_PRESET hidden)
+set(CMAKE_VISIBILITY_INLINES_HIDDEN ON)
+
+set_property(DIRECTORY PROPERTY COMPILE_OPTIONS
+    $<$<COMPILE_LANG_AND_ID:C,MSVC>:/W4 /wd4201>
+    $<$<COMPILE_LANG_AND_ID:C,Clang>:-Xclang -Wall -Wextra -Wno-microsoft-anon-tag -Wdocumentation>)
+
+add_subdirectory(external)
+
+if(NOT DEFINED ECM_MODULE_PATH)
+    find_package(ECM REQUIRED CONFIG)
+endif()
+list(APPEND CMAKE_MODULE_PATH ${ECM_MODULE_PATH})
+
+
+file(GLOB_RECURSE az_exampleshortname_src CONFIGURE_DEPENDS src/*.c)
+file(GLOB_RECURSE az_exampleshortname_test_src CONFIGURE_DEPENDS test/*.cpp)

```

Comment:
I have a vague recollection that globs in the build system are to be avoided; they're only updated on actual cmake runs, right?

Reply:
I recommend globs.

While this means that yes, adding a file to your disk won't automatically update your build, not using globs won't automatically update your build either! If you use globs, then you just need to trigger a reconfigure through any means (simply running `cmake .` in the build directory or hitting `Ctrl-S` on the CMakeLists.txt in visual studio). If you use an explicit sources list, you must edit the build file AND trigger a reconfigure.

Globs are also highly beneficial if someone wants to consume the project from an alternative buildsystem (bazel, pants, boost build, make, scons, msbuild, etc). This is because globs ensure that your build is _simple_ -- `*.c`. If you use an explicit list, it is extremely easy to accidentally introduce unwanted complexity. This also reduces cognitive load on anyone who needs to read the build files because there is only one source of truth: the files on disk.

Reply:
also, for most generators CONFIGURE_DEPENDS means adding a file will actually trigger a reconfigure of cmake.

Reply:
> If you use an explicit sources list, you must edit the build file AND trigger a reconfigure.

You lost me at "AND" here. Useful generators automatically reconfigure when the build is updated.

> If you use an explicit list, it is extremely easy to accidentally introduce unwanted complexity.

I suspect I'm biased by my experience which is heavily weighted toward header-only libraries in which every source file is a single-TU unit test.

> there is only one source of truth: the files on disk.

Which is all fine and dandy until you accidentally ship `tmp_hack.c`. ;)

In any case, I bow to the combination of Robert's expertise and Charlie's pointing out that my recollection predates `CONFIGURE_DEPENDS`. Feel free to resolve this comment.

Reply:
yeah so what convinced me is that with file(GLOB_RECURSE) you ensure (by construction) that your build system works by doing essentially `gcc src/*.c` which is a _really_ nice property to have. And the difference even between listing sources and globbing without `CONFIGURE_DEPENDS` is just that you `touch` the buildsystem instead of finding where to list the file and listing it.

The buildsystem doesn't have to like, inotify on your source directory either (hi bezel 🙌!), it just runs `ls -f` and compares output.